examples/vision/adamatch.py

```diff
@@ -0,0 +1,597 @@
+
+        gradients = tape.gradient(total_loss, self.model.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.model.trainable_variables))
+
+        self.loss_tracker.update_state(total_loss)
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+The authors introduce three improvements in the paper:
+
+* In AdaMatch, we perform two forward passes, and only one of them is respsonsible for
+updating the Batch Normalization statistics. This is done to account for distribution
+shifts in the target dataset. In the other forward pass, we only use the source sample,
+and the Batch Normalization layers are run in inference mode. Logits for the source
+samples (weakly and strongly augmented versions) from these two passes are slightly
+different from one another because of how Batch Normalization layers are run. Final
+logits for the source samples are computed by linearly interpolating between these two
+different pairs of logits. This induces a form of consistency regularization. This step
+is referred to as **random logit interpolation**.
+* **Distribution alignment** is used to align the source and target label distributions.
+This further helps the underlying model learn *domain-invariant representations*. In case
+of unsupervised domain adaptation, we don't have access to any labels of the target
+dataset. This is why pseudo labels are generated from the underlying model.
+* The underlying model generates pseudo-labels for the target samples. It's likely that
+the model would make faulty predictions. Those can propagate back as we make progress in
+the training, and hurt the overall performance. To compensate for that, we filter the
+high-confidence predictions based on a threshold (hence the use of `mask` inside
+`compute_loss_target()`). In AdaMatch, this threshold is relatively adjusted which is why
+it is called **relative confidence thresholding**.
+
+For more details on these methods and to know how each of them contribute please refer to
+[the paper](https://arxiv.org/abs/2106.04732).
+
+**About `compute_mu()`**:
+
+Rather than using a fixed scalar quantity, a varying scalar is used in AdaMatch. It
+denotes the weight of the loss contibuted by the target samples. Visually, the weight
+scheduler look like so:
+
+![](https://i.imgur.com/dG7i9uH.png)
+
+This scheduler increases the weight of the target domain loss from 0 to 1 for the first
+half of the training. Then it keeps that weight at 1 for the second half of the training.
+"""
+
+"""
+## Instantiate a Wide-ResNet-28-2
+
+The authors use a [WideResNet-28-2](https://arxiv.org/abs/1605.07146) for the dataset
+pairs we are using in this example. Most of the following code has been referred from
+[this script](https://github.com/asmith26/wide_resnets_keras/blob/master/main.py). Note
+that the following model has a scaling layer inside it that scales the pixel values to
+[0, 1].
+"""
+
+
+def wide_basic(n_input_plane, n_output_plane, stride):
```

Comment:
You can further simplify the code by making this function take a tensor and return a tensor (rather than a function)

Reply:
Done.